Orbitersdk/samples/ProjectApollo/src_sys/yaAGC/agc_engine.c

```diff
@@ -1905,6 +1905,15 @@ agc_engine (agc_t * State)
   //  }
 
   State->CycleCounter++;
+
```

Comment:
Are these changes also committed to virtualAGC? We should not start diverging from upstream again.

Reply:
> 
> 
> I missed the discussion leading to these changes so I'm not able to fully understand why these changes are needed. Could you please elaborate a bit on what you're trying to achieve? :)
> 
> There are quite a lot of changes to yaAGC. They should not be changed in NASSP but in VirtualAGC as that is our upstream. When the changes have been included there we can update our files to sync with their repo. Doing so also means these changes can be reviewed by more people.
> 
> Tip: It's best practice to always split your code into small commits detailing why that change was made. This not only makes it easier for others to look at your code but also for yourself to keep track of what you changed and to roll back to a previous state in case something breaks. It's a lot of work to go back and change this, so don't bother. Please do keep it in mind for next time though. :-)

The desired final state would be similar what you can see in the agcio branch (not this agcio2 ) in my repo. On agcio I have written a post in May where I summarized the goals of this development, you may read it here: https://www.orbiter-forum.com/threads/simulation-of-apollo-11-lunar-landing.37948/ . The main reason why I have to touch the AGC engine itself is that it should interact with most of the peripherials (CDUs, PIPAs, radars, etc.) in a synchronized way for proper simulation. The original goal was to create an accurate simulation of Apollo 11, but as a side effect it turned out that the accuracy of the PIPAs could be improved significantly by this. I also have some ideas on how the accuracy and fuel-efficiency of the DAP could be improved (or more precisely: to entirely utilize the "steady as rock" feature already provided by Colossus and Luminary softwares) but to be able to do so I'll also need this synchronous behaviour of AGC. We had a discussion on IRC with Mike and Indy (see the logs about from 2 weeks ago) on how to integrate all these stuff into the NASSP main branch. What came out of this was that the top priority is not to ruin any functionality which is available now in NASSP. My agcio branch currently fails this in two aspects: 1. it cannot raise hw alarms 2. not works well in multithreaded mode, but none of this seem to me a unsolvable issue, I just simply haven't taken care of them in the pervious development cycle. However making these changes compatible with VirtualAGC is just a nice-to-have thing at the moment. Anyway I've considered to make changes on VirtualAGC first and then migrate them into NASSP. The problem is that agc_engine.c in VirtualAGC has been _already_ diverged from the one in NASSP version, and I cannot see any reliable method to do this "NASSPsifying" after I make a change in VirtualAGC.

To make things clear, my roadmap for this transition is the following:
1. Increase the resolution of the scaler by two bits (or implement at all the scaler if you will), because that is the depth  required by peripherials. It could have been implemented outside of agc_engine.c but firstly I think as the scaler circuit is one of the core part of the real AGC hw so it's simulation should go into the emulator itself. Secondly, the agc_engine is doing some scaler mgmt already, if we repeat it outside of agc_engine, it will have impact on the performance. These steps are running 100k times in every simulated second, so in 100x acceleration they could have sensible impact.
2. Adding the properly implemented counter-register functionality to agc_engine. I intend to do this in a way that the AGC still remains compatible with the current (asynchronous) NASSP peripherials, but will be able accomodate synchronous devices as well. Will require some care here in case of outgoing registers, as they are behaving differently in the current emulation than on the real hw. In the worst case there will be some "#ifdef REAL_CDUERR_IMPL" parts in the code.
3. Hopefully after step 1-2 there will not be further need to modify agc_engine, and the synchronous and maybe more physics-oriented implementation of peripherials will happen in this step. I have done already this more-or-less completely for CDUs, the IMU gimbals, IMU coarse alignment, the PIPAs (I'm fairly proud of the in-depth but still surprisingly fast physical pendulum simulation there :) ), RR and optics angles and coarse alignment, LR and RR data readout, FDAI errors, CSM TVC/SIVB stuff, LM thrust drive. I haven't touched but they are working in their original (asynchronous) way: IMU Gyro Torquing, Telemetry Up and Downlink. To see what I intend to wire in, please refer to the commits of the agcio branch. In the best case  only the multithreading issue will inflict some changes till my codes make their way into agcio2.

So this PR corresponds to step 1.